File: packages/runar-rb/lib/runar/sdk/provider.rb
# frozen_string_literal: true

# Provider interface and MockProvider for testing.
#
# A Provider abstracts blockchain access: fetching UTXOs, looking up
# transactions, and broadcasting signed transactions to the network.

require_relative 'types'
require_relative 'errors'
require_relative 'bip143'

module Runar
  module SDK
    # Abstract base class for blockchain providers.
    #
    # Subclasses must implement all abstract methods. Raise NotImplementedError
    # from the default implementations keeps the intent explicit.
    class Provider
      # Fetch a Transaction by its txid.
      def get_transaction(_txid)
        raise NotImplementedError, "#{self.class}#get_transaction is not implemented"
      end

      # Fetch the raw transaction hex by its txid.
      def get_raw_transaction(_txid)
        raise NotImplementedError, "#{self.class}#get_raw_transaction is not implemented"
      end

      # Broadcast a raw transaction hex to the network.
      # Returns the txid of the broadcasted transaction.
      def broadcast(_raw_tx)
        raise NotImplementedError, "#{self.class}#broadcast is not implemented"
      end

      # Return all UTXOs for a given address.
      def get_utxos(_address)
        raise NotImplementedError, "#{self.class}#get_utxos is not implemented"
      end

      # Find a UTXO by its script hash (for stateful contract lookup).
      # Returns nil if not found.
      def get_contract_utxo(_script_hash)
        raise NotImplementedError, "#{self.class}#get_contract_utxo is not implemented"
      end

      # Return the network this provider is connected to (e.g. 'mainnet', 'testnet').
      def get_network
        raise NotImplementedError, "#{self.class}#get_network is not implemented"
      end

      # Return the current fee rate in satoshis per kilobyte.
      def get_fee_rate
        raise NotImplementedError, "#{self.class}#get_fee_rate is not implemented"
      end
    end

    # In-memory provider for unit tests and local development.
    #
    # Pre-populate with UTXOs and transactions, then inspect broadcasted
    # transactions after the fact.
    #
    #   provider = Runar::SDK::MockProvider.new
    #   provider.add_utxo('myAddress', Runar::SDK::Utxo.new(txid: 'abc...', ...))
    #   provider.get_utxos('myAddress') # => [<Utxo>]
    class MockProvider < Provider
      DEFAULT_FEE_RATE = 100

      # Shape of last_validation_report before any validating broadcast has run.
      EMPTY_VALIDATION_REPORT = {
        scripts_executed: 0,
        known_inputs: 0,
        total_inputs: 0,
        value_conserved: false
      }.freeze

      def initialize(network: 'testnet', validate_broadcasts: true)
        @transactions        = {}
        @utxos               = {}
        @contract_utxos      = {}
        @broadcasted_txs     = []
        @raw_transactions    = {}
        @broadcast_count     = 0
        @network             = network
        @fee_rate            = DEFAULT_FEE_RATE
        @validate_broadcasts = validate_broadcasts
        @known_outpoints     = {}
        @last_report         = EMPTY_VALIDATION_REPORT
      end

      # A MockProvider whose +broadcast+ never validates — the pre-Phase-A5
      # behaviour.
      #
      # FOR ALLOWLISTED SPECS ONLY: every spec file that calls this (or the
      # other opt-outs) must carry a matching entry in
      # +always_ack_allowlist.json+, enforced by
      # +spec/sdk/always_ack_allowlist_spec.rb+. Fund-path deploy/call specs
      # must not use it.
      def self.always_ack(network: 'testnet')
        new(network: network, validate_broadcasts: false)
      end

      # Turn the fail-closed +broadcast+ check on or off. Passing +false+ is an
      # allowlisted opt-out — see +MockProvider.always_ack+.
      def enable_broadcast_validation(enabled = true)
        @validate_broadcasts = enabled
      end

      # Restore the legacy always-ack +broadcast+. Allowlisted opt-out.
      def disable_broadcast_validation
        @validate_broadcasts = false
      end

      # Report from the most recent validating +broadcast+. Exposed so a spec
      # can assert its gate is NOT vacuous.
      #
      # +:scripts_executed+ is ALWAYS 0 in this tier and is present precisely so
      # that fact stays visible: Ruby ships no Bitcoin Script VM (see README,
      # "How fund-path tests fail closed in the Ruby tier").
      def last_validation_report
        @last_report
      end

      # Shorthand for +last_validation_report[:known_inputs]+ — the number of
      # spent outpoints this provider actually recognised and checked.
      def last_validated_input_count
        @last_report[:known_inputs]
      end

      # -- Mutation helpers ---------------------------------------------------

      # Register a transaction for later retrieval.
      def add_transaction(tx)
        @transactions[tx.txid] = tx
        Array(tx.outputs).each_with_index do |out, i|
          remember_outpoint(tx.txid, i, out.script, out.satoshis)
        end
      end

      # Register a UTXO under an address.
      def add_utxo(address, utxo)
        @utxos[address] ||= []
        @utxos[address] << utxo
        remember_outpoint(utxo.txid, utxo.output_index, utxo.script, utxo.satoshis)
      end

      # Register a UTXO under a script hash for stateful contract lookup.
      def add_contract_utxo(script_hash, utxo)
        @contract_utxos[script_hash] = utxo
        remember_outpoint(utxo.txid, utxo.output_index, utxo.script, utxo.satoshis)
      end

      # Override the fee rate (default: 100 sat/KB).
      def set_fee_rate(rate)
        @fee_rate = rate
      end

      # Return a copy of all raw transaction hexes that have been broadcasted.
      def get_broadcasted_txs
        @broadcasted_txs.dup
      end

      # -- Provider interface -------------------------------------------------

      def get_transaction(txid)
        tx = @transactions[txid]
        raise "MockProvider: transaction #{txid} not found" unless tx

        tx
      end

      def get_raw_transaction(txid)
        # Return auto-stored raw hex from a previous broadcast first.
        return @raw_transactions[txid] if @raw_transactions.key?(txid)

        tx = @transactions[txid]
        raise "MockProvider: transaction #{txid} not found" unless tx
        raise "MockProvider: transaction #{txid} has no raw hex" if tx.raw.to_s.empty?

        tx.raw
      end

      # Validate the transaction (unless validation has been opted out of) and
      # then record it, returning a deterministic fake txid.
      #
      # Fail-closed by default (testing-gap remediation Phase A5). This tier has
      # NO Bitcoin Script VM, so it makes no script-validity claim — see
      # +validate_broadcast!+ for exactly what it does and does not check.
      def broadcast(raw_tx)
        parsed = validate_broadcast!(raw_tx) if @validate_broadcasts

        @broadcasted_txs << raw_tx
        @broadcast_count += 1
        fake_txid = mock_hash64("mock-broadcast-#{@broadcast_count}-#{raw_tx[0, 16]}")
        # Auto-store raw hex so get_raw_transaction works without an explicit add_transaction call.
        @raw_transactions[fake_txid] = raw_tx
        # Register this tx's own outputs as known outpoints so a chained call
        # (spending the continuation this broadcast just created) is checkable.
        parsed&.fetch(:outputs)&.each_with_index do |out, i|
          remember_outpoint(fake_txid, i, out[:script].unpack1('H*'), out[:satoshis])
        end
        fake_txid
      end

      def get_utxos(address)
        utxos = Array(@utxos[address]).dup
        # DoS-bound: reject pathological scripts at the provider boundary.
        utxos.each do |u|
          next if u.script.nil? || u.script.empty?

          SDK.assert_script_hex_under_limit(
            u.script, SDK::MAX_SCRIPT_BYTES,
            "MockProvider.get_utxos(#{address})"
          )
        end
        utxos
      end

      def get_contract_utxo(script_hash)
        utxo = @contract_utxos[script_hash]
        if utxo && utxo.script && !utxo.script.empty?
          SDK.assert_script_hex_under_limit(
            utxo.script, SDK::MAX_SCRIPT_BYTES,
            "MockProvider.get_contract_utxo(#{script_hash})"
          )
        end
        utxo
      end

      def get_network
        @network
      end

      def get_fee_rate
        @fee_rate
      end

      private

      # Record an outpoint's script + value so broadcast validation can reason
      # about it.
      def remember_outpoint(txid, vout, script_hex, satoshis)
        return if txid.nil? || script_hex.nil? || script_hex.to_s.empty?

        @known_outpoints["#{txid}:#{vout}"] = { script: script_hex, satoshis: satoshis.to_i }
      end

      # Fail-closed broadcast validation (testing-gap remediation Phase A5).
      #
      # WHAT THIS CHECKS — and, just as importantly, what it does not.
      #
      # The Ruby tier ships no Bitcoin Script VM: there is no canonical upstream
      # BSV Ruby SDK to wrap, and project policy forbids hand-rolling an
      # interpreter (root CLAUDE.md, "Off-chain Script VM"). So this method
      # makes NO claim about signature or covenant validity. It applies the
      # checks that are genuinely available from the serialized bytes alone:
      #
      #   1. STRUCTURAL      — the payload must parse as a Bitcoin transaction.
      #   2. NON-VACUITY     — at least one spent outpoint must be known here,
      #                        so the gate can never pass by checking nothing.
      #   3. VALUE CONSERVE  — when every input is known, outputs <= inputs.
      #   4. SCRIPT-SIZE     — every output script stays under MAX_SCRIPT_BYTES.
      #
      # Script-level correctness for this tier is proven VERTICALLY instead:
      # absolute-hex pins against the peer tiers' goldens plus the on-chain
      # integration spends in integration/ruby.
      #
      # @raise [BroadcastRejected] when any check above fails
      # @return [Hash] the parsed transaction (for the caller to reuse)
      def validate_broadcast!(raw_tx)
        parsed = parse_broadcast_tx(raw_tx)

        known_inputs = 0
        total_known_in = 0
        all_inputs_known = true

        parsed[:inputs].each do |input|
          key = "#{input[:prev_txid].bytes.reverse.pack('C*').unpack1('H*')}:#{input[:prev_output_index]}"
          known = @known_outpoints[key]
          if known.nil?
            all_inputs_known = false
            next
          end
          known_inputs += 1
          total_known_in += known[:satoshis]
        end

        total_out = 0
        parsed[:outputs].each_with_index do |out, i|
          script_hex = out[:script].unpack1('H*')
          SDK.assert_script_hex_under_limit(
            script_hex, SDK::MAX_SCRIPT_BYTES, "MockProvider.broadcast output #{i}"
          )
          total_out += out[:satoshis]
        end

        @last_report = {
          scripts_executed: 0, # this tier has no Script VM — see the note above
          known_inputs: known_inputs,
          total_inputs: parsed[:inputs].length,
          value_conserved: all_inputs_known
        }.freeze

        if known_inputs.zero?
          raise BroadcastRejected,
                'MockProvider: refusing to broadcast — checked 0 of ' \
                "#{parsed[:inputs].length} input(s): no spent outpoint is known to this " \
                'provider, so validation would pass vacuously. Seed the spent outpoints via ' \
                'add_utxo/add_contract_utxo/add_transaction, or use MockProvider.always_ack ' \
                '(allowlisted) if this spec genuinely needs always-ack'
        end

        if all_inputs_known && total_out > total_known_in
          raise BroadcastRejected,
                "MockProvider: refusing to broadcast invalid transaction: underfunded: outputs " \
                "(#{total_out} sats) exceed known inputs (#{total_known_in} sats)"
        end

        parsed
      end

      def parse_broadcast_tx(raw_tx)
        hex = raw_tx.to_s
        raise BroadcastRejected, broadcast_parse_error(hex) unless hex.match?(/\A(?:[0-9a-fA-F]{2})+\z/)

        parsed = BIP143.parse_raw_tx([hex].pack('H*'))
        raise BroadcastRejected, broadcast_parse_error(hex) if parsed[:inputs].empty?

        parsed
      rescue ArgumentError, NoMethodError, TypeError
        raise BroadcastRejected, broadcast_parse_error(raw_tx.to_s)
      end

      def broadcast_parse_error(hex)
        'MockProvider: refusing to broadcast — payload is not a parseable Bitcoin transaction ' \
        "(#{hex.length / 2} byte(s)). A real node would reject it outright."
      end

      # Deterministic mock hash producing a 64-character hex string (like a txid).
      # Uses a simple FNV-inspired mix so the result is stable across Ruby versions.
      def mock_hash64(input)
        h0 = 0x6A09E667
        h1 = 0xBB67AE85
        h2 = 0x3C6EF372
        h3 = 0xA54FF53A
        mask32 = 0xFFFFFFFF

        input.each_char do |ch|
          c = ch.ord
          h0 = ((h0 ^ c) * 0x01000193) & mask32
          h1 = ((h1 ^ c) * 0x01000193) & mask32
          h2 = ((h2 ^ c) * 0x01000193) & mask32
          h3 = ((h3 ^ c) * 0x01000193) & mask32
        end

        parts = [h0, h1, h2, h3, h0 ^ h2, h1 ^ h3, h0 ^ h1, h2 ^ h3]
        parts.map { |p| format('%08x', p) }.join
      end
    end
  end
end

File: packages/runar-py/runar/sdk/provider.py
"""Provider interface and MockProvider for testing."""

from __future__ import annotations
from abc import ABC, abstractmethod
from runar.sdk.types import TransactionData, Utxo
from runar.sdk.errors import (
    BroadcastRejected,
    BroadcastValidationUnavailable,
    assert_script_hex_under_limit,
)
from runar.sdk.input_limits import MAX_SCRIPT_BYTES

_BSV_SDK_HINT = (
    "MockProvider's SCRIPT-EXECUTION layer needs the upstream bsv-sdk "
    "interpreter (bsv.script.spend.Spend), which is not installed.\n"
    "  Install it:  pip install bsv-sdk        (or: pip install 'runar[dev]')\n"
    "runar itself stays zero-dependency at RUNTIME; bsv-sdk is a TEST "
    "dependency, and CI installs it for the packages/runar-py job "
    "(.github/workflows/ci.yml: `pip install pytest slh-dsa bsv-sdk`).\n"
    "In CI ($CI set) its absence is a hard error, because CI must never lose "
    "the script layer silently. Outside CI the structural / non-vacuity / "
    "value-conservation layer still runs and the degradation is recorded as "
    "`script_vm_available: False` in last_validation_report — it is never "
    "silently reported as validated."
)


def _bsv_sdk_available() -> bool:
    try:
        import bsv.script.spend  # noqa: F401
        import bsv.transaction  # noqa: F401
    except ImportError:
        return False
    return True


class Provider(ABC):
    """Abstracts blockchain access for UTXO lookup and broadcast."""

    @abstractmethod
    def get_transaction(self, txid: str) -> TransactionData:
        """Fetch a transaction by its txid."""
        ...

    @abstractmethod
    def broadcast(self, tx) -> str:
        """Send a transaction to the network. Returns the txid.

        Accepts either a bsv-sdk Transaction object (calls tx.hex()) or a raw
        hex string for backward compatibility.
        """
        ...

    @abstractmethod
    def get_utxos(self, address: str) -> list[Utxo]:
        """Return all UTXOs for a given address."""
        ...

    @abstractmethod
    def get_contract_utxo(self, script_hash: str) -> Utxo | None:
        """Find a UTXO by its script hash (for stateful contract lookup)."""
        ...

    @abstractmethod
    def get_network(self) -> str:
        """Return the network this provider is connected to."""
        ...

    @abstractmethod
    def get_fee_rate(self) -> int:
        """Return the current fee rate in satoshis per KB (1000 bytes)."""
        ...

    @abstractmethod
    def get_raw_transaction(self, txid: str) -> str:
        """Fetch the raw transaction hex by its txid."""
        ...


_EMPTY_VALIDATION_REPORT = {
    'validated': 0,
    'unknown': 0,
    'unvalidatable': 0,
    'total': 0,
    'value_conserved': False,
    'script_vm_available': False,
}

# The ONE tolerated failure class, and only because this tier can currently
# validate NO Rúnar covenant at all — correct or broken alike.
#
# runar-py's OP_PUSH_TX preimage does not match the BIP-143 sighash bsv-sdk
# recomputes, so a stateful contract's covenant input always aborts at the
# covenant's OP_CHECKSIGVERIFY. This is pre-existing, independent of any SDK
# change, and already pinned by the xfail in
# ``tests/test_g1_raw_outputs_spend.py::test_send_to_script_covenant_input0_validates_through_spend``
# (its control shows a plain stateful Counter with no raw outputs fails
# identically). Such an input is recorded as UNVALIDATABLE — never as
# validated — so it can never satisfy the non-vacuity requirement on its own.
#
# ``tests/test_mock_broadcast_validation.py`` pins the incompatibility so this
# carve-out goes RED the moment it is fixed and can then be deleted.
_COVENANT_PUSHTX_MARKER = 'OP_CHECKSIGVERIFY'


class MockProvider(Provider):
    """In-memory provider for unit tests and local development.

    Broadcast validation is DEFAULT-ON (testing-gap remediation Phase A5) — see
    :meth:`broadcast` and README "How fund-path tests fail closed in the Python
    tier".
    """

    def __init__(self, network: str = 'testnet', validate_broadcasts: bool = True):
        self._transactions: dict[str, TransactionData] = {}
        self._raw_transactions: dict[str, str] = {}
        self._utxos: dict[str, list[Utxo]] = {}
        self._contract_utxos: dict[str, Utxo] = {}
        self._broadcasted_txs: list[str] = []
        self._network = network
        self._broadcast_count = 0
        self._fee_rate = 100
        self._validate_broadcasts = validate_broadcasts
        # "txid:vout" -> {'script': hex, 'satoshis': int}
        self._known_outpoints: dict[str, dict] = {}
        self._last_report = dict(_EMPTY_VALIDATION_REPORT)

    @classmethod
    def always_ack(cls, network: str = 'testnet') -> 'MockProvider':
        """A MockProvider whose :meth:`broadcast` never validates — the
        pre-Phase-A5 behaviour.

        FOR ALLOWLISTED TESTS ONLY: every test file that calls this (or the
        other opt-outs) must carry a matching entry in
        ``always_ack_allowlist.json``, enforced by
        ``tests/test_always_ack_allowlist.py``. Fund-path deploy/call tests
        must not use it.
        """
        return cls(network, validate_broadcasts=False)

    def enable_broadcast_validation(self, enabled: bool = True) -> None:
        """Turn the fail-closed :meth:`broadcast` check on or off. Passing
        ``False`` is an allowlisted opt-out — see :meth:`always_ack`."""
        self._validate_broadcasts = enabled

    def disable_broadcast_validation(self) -> None:
        """Restore the legacy always-ack :meth:`broadcast`. Allowlisted opt-out."""
        self._validate_broadcasts = False

    @property
    def last_validation_report(self) -> dict:
        """Report from the most recent validating :meth:`broadcast`. Exposed so
        a test can assert its gate is NOT vacuous."""
        return dict(self._last_report)

    @property
    def last_validated_input_count(self) -> int:
        """Number of inputs the most recent validating broadcast actually ran
        through the script interpreter."""
        return self._last_report['validated']

    def _remember_outpoint(self, txid, vout, script, satoshis) -> None:
        if not txid or not script:
            return
        self._known_outpoints[f'{txid}:{vout}'] = {
            'script': script,
            'satoshis': int(satoshis or 0),
        }

    def add_transaction(self, tx: TransactionData) -> None:
        self._transactions[tx.txid] = tx
        for i, out in enumerate(tx.outputs or []):
            self._remember_outpoint(tx.txid, i, out.script, out.satoshis)

    def add_utxo(self, address: str, utxo: Utxo) -> None:
        if address not in self._utxos:
            self._utxos[address] = []
        self._utxos[address].append(utxo)
        self._remember_outpoint(utxo.txid, utxo.output_index, utxo.script, utxo.satoshis)

    def add_contract_utxo(self, script_hash: str, utxo: Utxo) -> None:
        self._contract_utxos[script_hash] = utxo
        self._remember_outpoint(utxo.txid, utxo.output_index, utxo.script, utxo.satoshis)

    def get_broadcasted_txs(self) -> list[str]:
        return list(self._broadcasted_txs)

    def set_fee_rate(self, rate: int) -> None:
        self._fee_rate = rate

    # -- Provider interface --

    def get_transaction(self, txid: str) -> TransactionData:
        tx = self._transactions.get(txid)
        if tx is None:
            raise RuntimeError(f"MockProvider: transaction {txid} not found")
        return tx

    def broadcast(self, tx) -> str:
        """Validate the transaction (unless validation has been opted out of)
        and then record it, returning a deterministic fake txid.

        Fail-closed by default (testing-gap remediation Phase A5): every input
        whose outpoint this provider knows is executed by the upstream
        ``bsv-sdk`` ``Spend`` interpreter with full transaction context, outputs
        may not exceed known inputs, and a transaction none of whose inputs
        could be executed is REJECTED rather than waved through — a gate that
        validates nothing is worse than no gate.

        If ``bsv-sdk`` is missing, this raises
        :class:`~runar.sdk.errors.BroadcastValidationUnavailable` rather than
        skipping validation. See ``_BSV_SDK_HINT``.
        """
        # Accept either a bsv-sdk Transaction object or a raw hex string
        if isinstance(tx, str):
            raw_tx = tx
        else:
            raw_tx = tx.hex()

        parsed = None
        if self._validate_broadcasts:
            parsed = self._validate_broadcast(raw_tx)

        self._broadcasted_txs.append(raw_tx)
        # Auto-store raw hex for subsequent get_raw_transaction lookups
        self._broadcast_count += 1
        fake_txid = _mock_hash64(
            f"mock-broadcast-{self._broadcast_count}-{raw_tx[:16]}"
        )
        self._raw_transactions[fake_txid] = raw_tx
        # Register this tx's own outputs as known outpoints so a chained call
        # (spending the continuation this broadcast just created) is checkable.
        if parsed is not None:
            for i, out in enumerate(parsed['outputs']):
                self._remember_outpoint(fake_txid, i, out['script'].hex(), out['satoshis'] or 0)
        return fake_txid

    def _validate_broadcast(self, raw_tx: str):
        """Fail-closed broadcast validation. Returns the parsed transaction.

        TWO LAYERS, so the gate can never silently fail open:

        1. **Always on, stdlib only.** The payload must parse as a Bitcoin
           transaction (:func:`runar.sdk.oppushtx._parse_raw_tx`), at least one
           spent outpoint must be known here (non-vacuity), and — when every
           input is known — the outputs must not exceed the inputs.
        2. **Script execution via the upstream bsv-sdk ``Spend``**, when that
           optional dependency is installed. Its absence is NEVER a silent
           skip: it is recorded as ``script_vm_available: False`` in
           :attr:`last_validation_report`, and in CI (``$CI`` set) it is a hard
           :class:`BroadcastValidationUnavailable` error, because CI must never
           lose the script layer. See ``_BSV_SDK_HINT``.

        :raises BroadcastValidationUnavailable: bsv-sdk missing while in CI.
        :raises BroadcastRejected: the transaction is not parseable, an input's
            script fails, outputs exceed known inputs, or NOTHING could be
            checked at all (vacuous validation).
        """
        import os

        from runar.sdk.oppushtx import _parse_raw_tx

        script_vm = _bsv_sdk_available()
        if not script_vm and os.environ.get('CI'):
            raise BroadcastValidationUnavailable(_BSV_SDK_HINT)

        # --- layer 1: structural (stdlib only, always runs) ------------------
        try:
            raw_bytes = bytes.fromhex(raw_tx)
            structural = _parse_raw_tx(raw_bytes)
            if not structural['inputs'] or not structural['outputs']:
                raise ValueError('no inputs/outputs')
        except Exception as exc:
            raise BroadcastRejected(
                'MockProvider: refusing to broadcast — payload is not a parseable '
                f'Bitcoin transaction ({len(raw_tx) // 2} byte(s)): {exc}. '
                'A real node would reject it outright.'
            ) from exc

        unknown = 0
        all_inputs_known = True
        total_known_in = 0
        known_keys: list[str | None] = []

        for inp in structural['inputs']:
            key = f"{inp['prev_txid_bytes'][::-1].hex()}:{inp['prev_output_index']}"
            known = self._known_outpoints.get(key)
            if known is None:
                all_inputs_known = False
                unknown += 1
                known_keys.append(None)
            else:
                total_known_in += known['satoshis']
                known_keys.append(key)

        known_inputs = sum(1 for k in known_keys if k is not None)

        # --- layer 2: script execution (needs bsv-sdk) -----------------------
        parsed = None
        validated = 0
        unvalidatable = 0
        if script_vm:
            from bsv.script.script import Script as _BsvScript
            from bsv.script.spend import Spend as _BsvSpend
            from bsv.transaction import Transaction as _BsvTransaction

            parsed = _BsvTransaction.from_hex(raw_tx)
            for i, inp in enumerate(parsed.inputs):
                key = known_keys[i] if i < len(known_keys) else None
                if key is None:
                    continue
                known = self._known_outpoints[key]
                other_inputs = [v for j, v in enumerate(parsed.inputs) if j != i]
                try:
                    spend = _BsvSpend({
                        'sourceTXID': inp.source_txid,
                        'sourceOutputIndex': inp.source_output_index,
                        'sourceSatoshis': known['satoshis'],
                        'lockingScript': _BsvScript(known['script']),
                        'transactionVersion': parsed.version,
                        'otherInputs': other_inputs,
                        'outputs': parsed.outputs,
                        'inputIndex': i,
                        'unlockingScript': inp.unlocking_script,
                        'inputSequence': inp.sequence,
                        'lockTime': parsed.locktime,
                    })
                    ok = spend.validate()
                except Exception as exc:
                    if _COVENANT_PUSHTX_MARKER in str(exc):
                        unvalidatable += 1
                        continue
                    raise BroadcastRejected(
                        f'MockProvider: refusing to broadcast invalid transaction: '
                        f'input {i}: script REJECTED by bsv-sdk Spend: {exc}'
                    ) from exc
                if not ok:
                    raise BroadcastRejected(
                        'MockProvider: refusing to broadcast invalid transaction: '
                        f'input {i}: script evaluated to false'
                    )
                validated += 1

        self._last_report = {
            'validated': validated,
            'unknown': unknown,
            'unvalidatable': unvalidatable,
            'total': len(structural['inputs']),
            'value_conserved': all_inputs_known,
            'script_vm_available': script_vm,
        }

        # Non-vacuity: at least ONE real check must have run — a script really
        # executed, or (when every input's outpoint is known) the
        # value-conservation check below. If neither did, nothing was verified
        # and the ack would be a lie.
        if known_inputs == 0:
            raise BroadcastRejected(
                f'MockProvider: refusing to broadcast — NOTHING was checked '
                f"(0 of {len(structural['inputs'])} inputs recognised; unknown "
                f'outpoints: {unknown}), so validation would pass vacuously. Seed '
                'the spent outpoints via add_utxo/add_contract_utxo/add_transaction, '
                'or use MockProvider.always_ack() (allowlisted) if this test '
                'genuinely needs always-ack'
            )

        if all_inputs_known:
            total_out = sum(int(o['satoshis'] or 0) for o in structural['outputs'])
            if total_out > total_known_in:
                raise BroadcastRejected(
                    'MockProvider: refusing to broadcast invalid transaction: '
                    f'underfunded: outputs ({total_out} sats) exceed known inputs '
                    f'({total_known_in} sats)'
                )

        return structural

    def get_utxos(self, address: str) -> list[Utxo]:
        utxos = list(self._utxos.get(address, []))
        # DoS-bound: reject pathological scripts at the provider boundary.
        for u in utxos:
            if not u.script:
                continue
            assert_script_hex_under_limit(
                u.script, MAX_SCRIPT_BYTES,
                f"MockProvider.get_utxos({address})",
            )
        return utxos

    def get_contract_utxo(self, script_hash: str) -> Utxo | None:
        utxo = self._contract_utxos.get(script_hash)
        if utxo is not None and utxo.script:
            assert_script_hex_under_limit(
                utxo.script, MAX_SCRIPT_BYTES,
                f"MockProvider.get_contract_utxo({script_hash})",
            )
        return utxo

    def get_network(self) -> str:
        return self._network

    def get_fee_rate(self) -> int:
        return self._fee_rate

    def get_raw_transaction(self, txid: str) -> str:
        # Check auto-stored raw hex from broadcasts first
        if txid in self._raw_transactions:
            return self._raw_transactions[txid]
        tx = self._transactions.get(txid)
        if tx is None:
            raise RuntimeError(f"MockProvider: transaction {txid} not found")
        if not tx.raw:
            raise RuntimeError(f"MockProvider: transaction {txid} has no raw hex")
        return tx.raw


def _mock_hash64(input_str: str) -> str:
    """Deterministic mock hash producing a 64-char hex string (like a txid)."""
    h0 = 0x6A09E667
    h1 = 0xBB67AE85
    h2 = 0x3C6EF372
    h3 = 0xA54FF53A

    mask32 = 0xFFFFFFFF

    for ch in input_str:
        c = ord(ch)
        h0 = ((h0 ^ c) * 0x01000193) & mask32
        h1 = ((h1 ^ c) * 0x01000193) & mask32
        h2 = ((h2 ^ c) * 0x01000193) & mask32
        h3 = ((h3 ^ c) * 0x01000193) & mask32

    parts = [h0, h1, h2, h3, h0 ^ h2, h1 ^ h3, h0 ^ h1, h2 ^ h3]
    return ''.join(f'{p:08x}' for p in parts)
